File: python/0128longestConsecutiveSequence.py
class Solution:

    def longestConsecutive(self, nums: List[int]) -> int:
        """
        Time Complexity: O(N)
        Space Complexity: O(N)
        """
        if len(nums) < 2:
            return len(nums)

        seen = set(nums)
        longest = 1

        for i in range(len(nums)):
            if nums[i] - 1 in seen:
                continue

            inc = 1
            cur = 1
            while nums[i] + inc in seen:
                inc += 1
                cur += 1

            longest = max(cur, longest)

        return longest

    def longestConsecutive2(self, nums: List[int]) -> int:
        """
        Time Complexity: O(NlogN)
        Space Complexity: O(1)
        """
        if len(nums) < 2:
            return len(nums)
        # sort O(N * log N)
        nums.sort()

        longest = 1

        # start - i = len
        total = 1
        for i in range(len(nums) - 1):
            if nums[i] == nums[i + 1]:
                continue
            if nums[i] + 1 == nums[i + 1]:
                total += 1
            else:
                longest = max(total, longest)
                total = 1

        return max(total, longest)

File: javascript/0128longestConsecutiveSequence.js
/** Set
 * Time Complexity: O(N)
 * Space Complexity: O(N)
 */
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  let longest = 0;
  let set = new Set(nums);

  for (let n of set) {
    if (!set.has(n - 1)) {
      let cur = n;
      let streak = 1;

      while (set.has(cur + 1)) {
        cur++;
        streak++;
      }
      longest = Math.max(longest, streak);
    }
  }
  return longest;
};

/** Sorted
 * Time Complexity: O(Nlog(N))
 * Space Complexity: O(1) or O(N) if can't modify input
 */
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  if (nums.length < 2) return nums.length;
  nums.sort((a, b) => a - b);

  let longest = 1;
  let cur = 1;

  for (let i = 1; i < nums.length; i++) {
    if (nums[i - 1] !== nums[i]) {
      if (nums[i - 1] === nums[i] - 1) {
        cur += 1;
      } else {
        cur = 1;
      }
      if (cur > longest) {
        longest = cur;
      }
    }
  }
  return longest;
};
